jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ResourceService.java

```diff
@@ -728,55 +736,55 @@ protected void writeHttpContent(Request request, Response response, Callback cal
 
     protected void putHeaders(Response response, HttpContent content, long contentLength)
     {
-        // TODO it is very inefficient to do many put's to a HttpFields, as each put is a full iteration.
-        //      it might be better remove headers en masse and then just add the extras:
-        // NOTE: If these headers come from a Servlet Filter we shouldn't override them here.
-//        headers.remove(EnumSet.of(
-//            HttpHeader.LAST_MODIFIED,
-//            HttpHeader.CONTENT_LENGTH,
-//            HttpHeader.CONTENT_TYPE,
-//            HttpHeader.CONTENT_ENCODING,
-//            HttpHeader.ETAG,
-//            HttpHeader.ACCEPT_RANGES,
-//            HttpHeader.CACHE_CONTROL
-//            ));
-//        HttpField lm = content.getLastModified();
-//        if (lm != null)
-//            headers.add(lm);
-//        etc.
+        HttpFields.Mutable headers = response.getHeaders();
 
+        // Existing etags have priority over content etags (often set by compression handler)
+        if (_etags && !headers.contains(HttpHeader.ETAG))
+        {
+            HttpField et = content.getETag();
+            if (et != null)
+                headers.add(et);
+        }
+
+        // Existing content encoding is kept only if not set for content.
+        HttpField ce = content.getContentEncoding();
+        if (ce != null)
+            headers.put(ce);
+
+        // Remove content headers and re-add if we have them
+        headers.remove(CONTENT_HEADERS);
```

Comment:
Is it ok to remove these headers if they have been added by the application?

Reply:
We were doing that before with put